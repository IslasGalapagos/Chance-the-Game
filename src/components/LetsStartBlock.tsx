/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import {StyledWrapper, StyledButton} from './LetsStartBlock.styles';

export interface Props {
  name: string;
  onSubmit: () => void;
}

class LetsStartBlock extends React.PureComponent<Props, {}> {
  render() {
    const {name, onSubmit} = this.props;

    return (
      <StyledWrapper>
        Hello,&#32;
        <span>{name}</span>
        .&#32;
        <StyledButton type="button" onClick={onSubmit}>
          Let's play!
        </StyledButton>
      </StyledWrapper>
    );
  }
}

export default LetsStartBlock;
