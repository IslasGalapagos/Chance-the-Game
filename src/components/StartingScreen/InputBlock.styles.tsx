/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import styled, {keyframes} from 'react-emotion';
import {fontStyles, margin} from './styles';

interface StyledProps {
  isEmpty: boolean;
}

export const StyledInputWrapper = styled('div')<StyledProps>`
  ${() => margin} input {
    border-bottom-color: ${({isEmpty}) => (isEmpty ? '#f55662' : '')};
  }
`;

const animatedLabelKeyframes = keyframes`
  from {
    left: 0;
  }

  50% {
    left: 4px;
  }

  to {
    left: 0;
  }
`;

export const AnimatedLabel = styled('label')`
  ${() => fontStyles} position: relative;
  animation: ${animatedLabelKeyframes} 0.7s ease infinite;
`;

export const StyledInput = styled('input')`
  ${() => fontStyles} width: 241px;
  margin-left: 20px;

  background-color: transparent;
  border: none;
  border-bottom: 1px solid #ccc;
  outline: none;
`;
