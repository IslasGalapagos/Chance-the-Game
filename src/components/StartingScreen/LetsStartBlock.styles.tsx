/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import styled, {css, keyframes} from 'react-emotion';
import {fontStyles} from './styles';

const animatedWrapperKeyframes = keyframes`
  from {
    transform: scale(0);
  }
  to {
    transform: scale(1);
  }
`;

const animatedWrapper = css`
  animation: ${animatedWrapperKeyframes} 0.1s ease;
`;

export const StyledWrapper = styled('div')`
  ${() => fontStyles} display: inline-block;

  ${() => animatedWrapper};
`;

export const StyledButton = styled('button')`
  ${() => fontStyles};
  font-style: italic;

  padding: 0;

  background-color: transparent;
  color: #3e3cae;
  border: none;
  border-bottom: 1px solid #3e3cae;
  outline: none;

  cursor: pointer;
`;
