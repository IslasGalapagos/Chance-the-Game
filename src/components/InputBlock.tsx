import * as React from 'react';
import styled, {keyframes} from 'react-emotion';
import {State as StartingScrennState, fontStyles, margin} from './StartingScreen';

interface StyledProps {
  isEmpty: boolean;
}

export const StyledInputWrapper = styled('div')<StyledProps>`
  input {
    border-bottom-color: ${({isEmpty}) => (isEmpty ? '#f55662' : '')};
  }
`;

const animatedLabelKeyframes = keyframes`
  from {
    left: 0;
  }

  50% {
    left: 4px;
  }

  to {
    left: 0;
  }
`;

const AnimatedLabel = styled('label')`
  position: relative;
  animation: ${animatedLabelKeyframes} 0.7s ease infinite;
`;

export const StyledInput = styled('input')`
  width: 241px;
  margin-left: 20px;

  background-color: transparent;
  border: none;
  border-bottom: 1px solid #ccc;
  outline: none;
`;

export interface Props extends Pick<StartingScrennState, 'name'> {
  isEmpty: boolean;
  onInput: (event: React.FormEvent<HTMLInputElement>) => void;
  focus: (isFocused: boolean) => void;
}

class InputBlock extends React.PureComponent<Props, {}> {
  constructor(props: any) {
    super(props);

    this.focus = this.focus.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
  }

  focus(event: React.FocusEvent<HTMLInputElement>): void {
    this.props.focus(event.type === 'focus');
  }

  onKeyPress(event: React.KeyboardEvent<HTMLInputElement>): void {
    if (event.which === 13) {
      event.currentTarget.blur();
    }
  }

  render() {
    const {isEmpty, name, onInput} = this.props;

    return (
      <StyledInputWrapper className={margin} isEmpty={isEmpty}>
        <AnimatedLabel className={fontStyles} htmlFor="name">
          &#8680;
        </AnimatedLabel>
        <StyledInput
          className={fontStyles}
          id="name"
          type="text"
          autoComplete="off"
          value={name}
          onChange={onInput}
          onFocus={this.focus}
          onBlur={this.focus}
          onKeyPress={this.onKeyPress}
          autoFocus={true}
        />
      </StyledInputWrapper>
    );
  }
}

export default InputBlock;
