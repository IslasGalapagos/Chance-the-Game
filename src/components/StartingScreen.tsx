/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import {StyledWrapper, StyledHeader} from './StartingScreen.styles';
import LetsStartBlock from './LetsStartBlock';
import InputBlock from './InputBlock';

export const enum Header {
  Question = "What's your name?",
  Asking = 'Tell me.'
}

export interface State {
  readonly header: Header;
  readonly name: string;
  readonly inputIsFocused: boolean;
}

class StartingScreen extends React.PureComponent<{}, State> {
  readonly state: State;

  constructor(props: any) {
    super(props);

    this.state = {
      header: Header.Question,
      name: '',
      inputIsFocused: true
    };

    this.onInput = this.onInput.bind(this);
    this.focus = this.focus.bind(this);
  }

  onInput(event: React.FormEvent<HTMLInputElement>): void {
    const {value} = event.currentTarget;
    const {name, header} = this.state;

    let newHeader: Header | null = null;

    if (name.length === 0 && value.length > 0 && header === Header.Asking) {
      newHeader = Header.Question;
    }

    if (name.length > 0 && value.length === 0 && header === Header.Question) {
      newHeader = Header.Asking;
    }

    this.setState({
      name: name !== value ? value : name,
      header: newHeader || header
    });
  }

  focus(isFocused: boolean): void {
    const {inputIsFocused, header} = this.state;

    let newHeader: Header | null = null;

    if (inputIsFocused && header === Header.Asking) {
      newHeader = Header.Question;
    }

    this.setState({
      inputIsFocused: isFocused,
      header: newHeader || header
    });
  }

  render() {
    const {header, name, inputIsFocused} = this.state;

    return (
      <StyledWrapper hidden>
        <StyledHeader>
          {header}
        </StyledHeader>
        <InputBlock
          isEmpty={name.length === 0}
          name={name}
          onInput={this.onInput}
          focus={this.focus}
        />
        {!!name.length && !inputIsFocused && <LetsStartBlock name={name} />}
      </StyledWrapper>
    );
  }
}

export default StartingScreen;
