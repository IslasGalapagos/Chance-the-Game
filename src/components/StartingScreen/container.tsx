/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import {connect} from 'react-redux';
import {Dispatch} from 'redux';
import {State as ReducerState} from '../../store/reducer';
import {setName} from '../../store/actions';
import StartingScreen from '.';

export interface Props {
  name: string;
  setName: (name: string) => Dispatch<ReturnType<typeof setName>>;
}

export class Wrapper extends React.PureComponent<Props, {}> {
  static readonly displayName = 'StartingScreenContainer';

  constructor(props: any) {
    super(props);
  }

  render() {
    const {props} = this;

    if (props.name.length) {
      return null;
    }

    return <StartingScreen {...props} />;
  }
}

const mapStateToProps = (state: ReducerState) => ({
  name: state.name
});

const mapDispatchToProps = (
  dispatch: Dispatch<ReturnType<typeof setName>>
) => ({
  setName: (name: string) => dispatch(setName(name))
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Wrapper);
