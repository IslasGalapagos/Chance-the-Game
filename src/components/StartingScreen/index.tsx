/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import * as React from 'react';
import {StyledWrapper, StyledHeader} from './styles';
import LetsStartBlock from './LetsStartBlock';
import InputBlock from './InputBlock';
import {Props} from './container';

export const enum Header {
  Question = "What's your name?",
  Asking = 'Tell me.'
}

export interface State {
  readonly header: Header;
  readonly inputVal: string;
  readonly inputIsFocused: boolean;
}

export class StartingScreen extends React.PureComponent<Props, State> {
  readonly state: State;

  constructor(props: any) {
    super(props);

    this.state = {
      header: Header.Question,
      inputVal: '',
      inputIsFocused: true
    };

    this.onInput = this.onInput.bind(this);
    this.focus = this.focus.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  onInput(event: React.FormEvent<HTMLInputElement>): void {
    const {value} = event.currentTarget;
    const {inputVal, header} = this.state;

    let newHeader: Header | null = null;

    if (inputVal.length === 0 && value.length > 0 && header === Header.Asking) {
      newHeader = Header.Question;
    }

    if (
      inputVal.length > 0 &&
      value.length === 0 &&
      header === Header.Question
    ) {
      newHeader = Header.Asking;
    }

    this.setState({
      inputVal: inputVal !== value ? value : inputVal,
      header: newHeader || header
    });
  }

  focus(isFocused: boolean): void {
    const {inputIsFocused, header} = this.state;

    let newHeader: Header | null = null;

    if (inputIsFocused && header === Header.Asking) {
      newHeader = Header.Question;
    }

    this.setState({
      inputIsFocused: isFocused,
      header: newHeader || header
    });
  }

  onSubmit() {
    this.props.setName(this.state.inputVal);
  }

  render() {
    const {header, inputVal, inputIsFocused} = this.state;

    return (
      <StyledWrapper>
        <StyledHeader>{header}</StyledHeader>
        <InputBlock
          isEmpty={inputVal.length === 0}
          inputVal={inputVal}
          onInput={this.onInput}
          focus={this.focus}
        />
        {!!inputVal.length &&
          !inputIsFocused && (
            <LetsStartBlock name={inputVal} onSubmit={this.onSubmit} />
          )}
      </StyledWrapper>
    );
  }
}

export default StartingScreen;
