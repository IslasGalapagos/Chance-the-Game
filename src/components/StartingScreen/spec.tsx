/*
  This Source Code Form is subject to the terms of the Mozilla Public
  License, v. 2.0. If a copy of the MPL was not distributed with this
  file, You can obtain one at http://mozilla.org/MPL/2.0/.

  Copyright Â© 2018 Evgeny Sysoletin. All rights reserved.
*/

import {shallow, mount} from 'enzyme';
import * as testRenderer from 'react-test-renderer';
import * as React from 'react';
import StartingScreen, {Header} from '.';
import {Props} from './container';
import LetsStartBlock from './LetsStartBlock';
import InputBlock from './InputBlock';

const props: Props = {
  name: '',
  setName: jest.fn()
};

describe('<StartingScreen/>', () => {
  it('renders correctly', () => {
    const treeAndStyles = testRenderer
      .create(<StartingScreen {...props} />)
      .toJSON();
    expect(treeAndStyles).toMatchSnapshot();
  });

  // STATE
  it(`sets '' to state.inputVal by default`, () => {
    const component = shallow(<StartingScreen {...props} />);
    expect(component.state('inputVal')).toEqual('');
  });

  // METHODS
  it('has onSubmit that calls props.setName with state.inputVal', () => {
    const component = shallow(<StartingScreen {...props} />);
    component.setState({inputVal: 'Diophantus'});
    const {onSubmit} = component.instance() as StartingScreen;
    onSubmit();
    expect(props.setName).toBeCalledTimes(1);
    expect(props.setName).toBeCalledWith('Diophantus');

    (props.setName as any).mockClear();
  });

  // HEADER
  it(`changes text of header-question when state.inputIsFocused is true
      & state.inputVal changes to empty string`, () => {
    const component = mount(<StartingScreen {...props} />);
    const header = component.find('h1');
    const instance = component.instance() as StartingScreen;
    const onInput = instance.onInput as (
      event: {currentTarget: {value: string}}
    ) => void;

    expect(header.text()).toEqual(Header.Question);
    instance.focus(true);
    expect(header.text()).toEqual(Header.Question);

    onInput({currentTarget: {value: 'Diophantus'}});
    expect(header.text()).toEqual(Header.Question);
    onInput({currentTarget: {value: ''}});
    expect(header.text()).toEqual(Header.Asking);

    instance.focus(false);
    expect(header.text()).toEqual(Header.Question);
  });

  // BUTTON
  it('hides button by default, shows when inputVal.length is not equals 0 & state.inputIsFocused is false', () => {
    const component = shallow(<StartingScreen {...props} />);
    expect(component.find(LetsStartBlock).length).toEqual(0);
    component.setState({inputVal: 'Diophantus'});
    expect(component.find(LetsStartBlock).length).toEqual(0);
    component.setState({inputIsFocused: false});
    expect(component.find(LetsStartBlock).length).toEqual(1);
    component.setState({inputIsFocused: true});
    component.setState({inputVal: ''});
    expect(component.find(LetsStartBlock).length).toEqual(0);
  });

  // INPUT
  it('sets isEmpty prop to input wrapper when inputVal.length is equals 0', () => {
    const component = shallow(<StartingScreen {...props} />);
    expect(component.find(InputBlock).prop('isEmpty')).toEqual(true);
    component.setState({inputVal: 'Diophantus'});
    expect(component.find(InputBlock).prop('isEmpty')).toEqual(false);
    component.setState({inputVal: ''});
    expect(component.find(InputBlock).prop('isEmpty')).toEqual(true);
  });

  it(`sets state.inputVal to InputBlock's prop inputVal`, () => {
    const component = shallow(<StartingScreen {...props} />);
    expect(component.find(InputBlock).prop('inputVal')).toEqual('');
    component.setState({inputVal: 'Diophantus'});
    expect(component.find(InputBlock).prop('inputVal')).toEqual('Diophantus');
  });

  // LETSSTART
  it(`sets state.inputVal to LetsStartBlock's prop name`, () => {
    const component = shallow(<StartingScreen {...props} />);
    component.setState({inputVal: 'Diophantus'});
    component.setState({inputIsFocused: false});
    expect(component.find(LetsStartBlock).prop('name')).toEqual('Diophantus');
    component.setState({inputVal: 'Pythagoras'});
    expect(component.find(LetsStartBlock).prop('name')).toEqual('Pythagoras');
  });

  it(`sets onSubmit to LetsStartBlock's onSubmit prop`, () => {
    const component = shallow(<StartingScreen {...props} />);
    component.setState({inputVal: 'Diophantus'});
    component.setState({inputIsFocused: false});
    const {onSubmit} = component.instance() as StartingScreen;
    expect(component.find(LetsStartBlock).prop('onSubmit')).toEqual(onSubmit);
  });
});
