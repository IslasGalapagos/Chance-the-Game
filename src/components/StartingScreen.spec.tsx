import {shallow, mount} from 'enzyme';
import * as testRenderer from 'react-test-renderer';
import * as React from 'react';
import StartingScreen, {Header} from './StartingScreen';
import LetsStartBlock from './LetsStartBlock';
import InputBlock from './InputBlock';

describe('<StartingScreen/>', () => {
  it('renders correctly', () => {
    const treeAndStyles = testRenderer.create(<StartingScreen />).toJSON();
    expect(treeAndStyles).toMatchSnapshot();
  });

  it('has initial state with name property which equals empty string', () => {
    const component = shallow(<StartingScreen />);
    expect(component.state('name')).toEqual('');
  });

  it('hides button by default, shows when name.length not equals 0 & state.inputIsFocused is false', () => {
    const component = shallow(<StartingScreen />);
    expect(component.find(LetsStartBlock).length).toEqual(0);
    component.setState({name: 'Diophantus'});
    expect(component.find(LetsStartBlock).length).toEqual(0);
    component.setState({inputIsFocused: false});
    expect(component.find(LetsStartBlock).length).toEqual(1);
    component.setState({inputIsFocused: true});
    component.setState({name: ''});
    expect(component.find(LetsStartBlock).length).toEqual(0);
  });

  it(`changes text of header-question when state.inputIsFocused is true & state.name changes to empty string`, () => {
    const component = mount(<StartingScreen />);
    const header = component.find('h1');
    const instance = component.instance() as StartingScreen;
    const onInput = instance.onInput as (
      event: {currentTarget: {value: string}}
    ) => void;

    expect(header.text()).toEqual(Header.Question);
    instance.focus(true);
    expect(header.text()).toEqual(Header.Question);

    onInput({currentTarget: {value: 'Diophantus'}});
    expect(header.text()).toEqual(Header.Question);
    onInput({currentTarget: {value: ''}});
    expect(header.text()).toEqual(Header.Asking);

    instance.focus(false);
    expect(header.text()).toEqual(Header.Question);
  });

  it('sets isEmpty prop to input wrapper when name.length equals 0', () => {
    const component = shallow(<StartingScreen />);
    expect(component.find(InputBlock).prop('isEmpty')).toEqual(true);
    component.setState({name: 'Diophantus'});
    expect(component.find(InputBlock).prop('isEmpty')).toEqual(false);
    component.setState({name: ''});
    expect(component.find(InputBlock).prop('isEmpty')).toEqual(true);
  });

  it(`sets state.name to InputBlock's prop name`, () => {
    const component = shallow(<StartingScreen />);
    expect(component.find(InputBlock).prop('name')).toEqual('');
    component.setState({name: 'Diophantus'});
    expect(component.find(InputBlock).prop('name')).toEqual('Diophantus');
  });

  it(`sets state.name to LetsStartBlock's prop name`, () => {
    const component = shallow(<StartingScreen />);
    component.setState({name: 'Diophantus'});
    component.setState({inputIsFocused: false});
    expect(component.find(LetsStartBlock).prop('name')).toEqual('Diophantus');
    component.setState({name: 'Pythagoras'});
    expect(component.find(LetsStartBlock).prop('name')).toEqual('Pythagoras');
  });
});
