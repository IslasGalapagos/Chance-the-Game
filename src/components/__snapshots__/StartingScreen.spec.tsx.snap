// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StartingScreen/> renders correctly 1`] = `
.emotion-4 {
  position: absolute;
  top: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  height: 100%;
  padding: 35px;
  box-sizing: border-box;
  background-color: #eafff2;
}

.emotion-0 {
  font-weight: normal;
  font-family: Arial;
  font-size: 35px;
  color: #333;
  margin: 0 0 25px 0;
}

.emotion-3 {
  margin: 0 0 25px 0;
}

.emotion-3 input {
  border-bottom-color: #f55662;
}

.emotion-1 {
  position: relative;
  -webkit-animation: animation-e59sno 0.7s ease infinite;
  animation: animation-e59sno 0.7s ease infinite;
  font-family: Arial;
  font-size: 35px;
  color: #333;
}

.emotion-2 {
  width: 241px;
  margin-left: 20px;
  background-color: transparent;
  border: none;
  border-bottom: 1px solid #ccc;
  outline: none;
  font-family: Arial;
  font-size: 35px;
  color: #333;
}

<div
  className="emotion-4"
>
  <h1
    className="emotion-0"
  >
    What's your name?
  </h1>
  <div
    className="emotion-3"
  >
    <label
      className="emotion-1"
      htmlFor="name"
    >
      â‡¨
    </label>
    <input
      autoComplete="off"
      autoFocus={true}
      className="emotion-2"
      id="name"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyPress={[Function]}
      type="text"
      value=""
    />
  </div>
</div>
`;
